core/trino-main/src/main/java/io/trino/sql/planner/ExpressionInterpreter.java

```diff
@@ -159,7 +159,8 @@
 
 public class ExpressionInterpreter
 {
-    private static final Map<InListExpression, Set<?>> inListCache = new WeakHashMap<>();
+    private static final Map<InListExpression, Set<?>> inListLiteralCache = new WeakHashMap<>();
+    private static final Map<InListExpression, List<Object>> inListExpressionCache = new WeakHashMap<>();
```

Comment:
I don't think we need `inListExpressionCache` since you introduce `optimizationCache` in the next commit. Within `optimizationCache` you can keep optimizer version of `InListExpression`